### PR TITLE
Support default data type

### PR DESCRIPTION
**Pull request by sjwhalen**
* Add default_data_type param to show_output_to_df
* Add dataframe_assert_equal and its test

### CODE REVIEW

**Review by josh803316 — APPROVED** [review]
@sjwhalen Should we add someone from Eng with a Python background, maybe Chris? I haven't used it in years :)

**Comment by sjwhalen** [comment]
@josh803316: That sounds good. I'll see if @jdavenport can take a look, as he has contributed to the library before.

File: dataframe_show_reader/dataframe_assert_equal.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    """Assert that two DataFrames contain the same data.
+    :param expected: The expected DataFrame
+    :param actual: The actual DataFrame
+    :param show_output_on_failure: If the DataFrames are not equal, show
+    the DataFrame schema or data (depending on where the mismatch is) in order
+    to help debugging.
+    :return: None
+    """
+    if expected is None:
+        assert actual is None, \
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_assert_equal.py]: If show_output_on_failure is True, should we display the non-None dataframe?

File: dataframe_show_reader/dataframe_assert_equal.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    else:
+        assert actual is not None, \
+            'The actual DataFrame is None, but the expected DataFrame is not.'
+
+    expected.persist()
+    actual.persist()
+
+    # When comparing the schemas, ignore the metadata, which can include
+    # comments.
+    actual_schema_without_metadata = StructType(
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_assert_equal.py]: Is it possible that the expected schema has metadata with comments that we would want to ignore, too?
Also, does the order (alphabetically by column name) of the schema matter?  Perhaps not, but i know i've hit this issue before in a union of 2 dataframes.

File: dataframe_show_reader/dataframe_assert_equal.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        [StructField(f.name, f.dataType, f.nullable) for f in actual.schema]
+    )
+    is_schema_match = expected.schema == actual_schema_without_metadata
+    if not is_schema_match:
+        if show_output_on_failure:
+            print('Expected DataFrame schema:')
+            expected.printSchema()
+            print('Actual DataFrame schema:')
+            actual.printSchema()
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_assert_equal.py]: I wonder if it's easier to compare the schemas visually by doing a show(1, False) on both.  or show(0, False) if that actually prints anything.

File: dataframe_show_reader/dataframe_assert_equal.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        assert is_schema_match is True, 'The DataFrame schemas differ.'
+
+    is_match = (0 == (expected.count() - actual.count()) ==
+                expected.subtract(actual).count() ==
+                actual.subtract(expected).count())
+
+    if not is_match:
+        if show_output_on_failure:
+            print('Expected DataFrame:')
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_assert_equal.py]: Maybe change the message to 'Expected DataFrame (first 20 rows):'

File: dataframe_show_reader/dataframe_show_reader.py [hunk 5]
```diff
@@ -59,10 +62,13 @@ def show_output_to_df(show_output: str, spark_session: SparkSession):
     :param show_output: A string that resembles the output of a call to
            DataFrame.show()
     :param spark_session: A SparkSession used to create the new DataFrame instance
+    :param default_data_type: The default data type that will be used for all
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_show_reader.py:65]: Nice.

File: dataframe_show_reader/dataframe_show_reader.py [hunk 6]
```diff
@@ -76,19 +82,29 @@ def show_output_to_df(show_output: str, spark_session: SparkSession):
...
+
+        if types is None:
+            types = [default_data_type] * len(column_names)
+
+        _cast_types(values, types)
         row_dict = dict(zip(column_names, values))
         rows.append(Row(**row_dict))
     # Return a DataFrame with the columns in the original order:
+    if types is None:
+        types = [default_data_type] * len(column_names)
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_show_reader.py:109]: Can you add a comment about when this condition can occur?  It's hard to follow.

File: dataframe_show_reader/dataframe_show_reader.py [hunk 7]
```diff
@@ -128,7 +144,8 @@ def save_df_as_table(table_name: str,
     :param partition_cols: Optional list of columns on which to partition the table
     :return: None
     """
-    df.write.partitionBy(partition_cols).saveAsTable(table_name, mode='Overwrite')
+    df.repartition(1).write.partitionBy(partition_cols) \
```
Comment (jdavenport) [inline dataframe_show_reader/dataframe_show_reader.py:152]: Good idea.

File: tests/test_dataframe_show_reader.py [hunk 8]
```diff
@@ -122,13 +191,32 @@ def test_show_output_to_df_when_values_are_null(
...
+        assert None is row['bool_column']
+
+    def test_show_output_to_df_when_values_are_empty_and_data_types_are_specified(
+            self,
+            spark_session: SparkSession
+    ):
+        rows = show_output_to_df("""
+        +-------------+-------------------+
+        |string_column|default_type_column|
+        [string       |                   |
```
Comment (jdavenport) [inline tests/test_dataframe_show_reader.py:240]: What do we get if the data type is 'int' and the value is not set?  I think you added this support, but maybe not.

**Review by jdavenport — CHANGES_REQUESTED** [review]
Looks good, steve.  a few comments.

File: dataframe_show_reader/dataframe_assert_equal.py (hunk 1 above)
Comment (sjwhalen) [reply dataframe_show_reader/dataframe_assert_equal.py]: I'm thinking that the error message (`The [x] DataFrame is None, but the [y] DataFrame is not.`) contains the important information and that the show output would be a distraction, but I could also be convinced that a strict literal interpretation of `show_output_on_failure` is appropriate. Although, if there is a schema mismatch, we call `printSchema` rather than `show`, which I think is the most useful output there, so maybe we should change the parameter name to `verbose`. I think I favor changing the parameter name to `verbose` and leaving the rest the same. Any thoughts?

Comment (jdavenport) [reply dataframe_show_reader/dataframe_assert_equal.py]: Yes, i like the `verbose` suggestion and leaving it as is.

File: dataframe_show_reader/dataframe_assert_equal.py (hunk 2 above)
Comment (sjwhalen) [reply dataframe_show_reader/dataframe_assert_equal.py]: Yes, that is a good point. I added code to remove metadata from the expected schema as well. I also added the optional param `ignore_schema_metadata: bool = True` in case someone does actually have a desire to check that the metadata matches.

The order of the columns does matter. I have added this test to confirm and document that fact:
```
def test_assert_equal_when_column_order_is_different(
        expected_df, spark_session: SparkSession):
    actual_df = show_output_to_df("""
    +-----+-----+
    |col_b|col_a|
    +-----+-----+
    |1b   |1a   |
    |2b   |2a   |
    +-----+-----+
    """, spark_session)

    with raises(AssertionError) as exception_info:
        assert_equal(expected_df, actual_df)
    assert 'The DataFrame schemas differ.' == str(exception_info.value)
```

File: dataframe_show_reader/dataframe_assert_equal.py (hunk 3 above)
Comment (sjwhalen) [reply dataframe_show_reader/dataframe_assert_equal.py]: There are some sorts of differences that `show()` would not demonstrate that could nonetheless be important. I have added this comment to describe this situation:
```
            # Print the schema to help identify subtle cases such as one DF
            # containing an INT and the other containing a BIGINT, which can be
            # an issue if we try to write a DF containing a BIGINT into a
            # previously existing Hive table defined to contain an INT.
```
This also reminded me to update `test_assert_equal_when_data_types_do_not_match` to test `INT` vs. `BIGINT`, which I had been meaning to do.

File: dataframe_show_reader/dataframe_assert_equal.py (hunk 4 above)
Comment (sjwhalen) [reply dataframe_show_reader/dataframe_assert_equal.py]: I have confirmed that anytime a DataFrame contains more rows than the `show` statement is going to print (either due to specifying the `n` parameter or using the Spark default `n` value of 20), the show output includes the message `only showing top <n> rows`.

File: dataframe_show_reader/dataframe_show_reader.py (hunk 6 above)
Comment (sjwhalen) [reply dataframe_show_reader/dataframe_show_reader.py:109]: Good idea. I wouldn't have though of this condition if a test case hadn't caught it. I have added this comment:
```
        # This can happen if data types are not specified and no data rows are
        # provided.
```

File: tests/test_dataframe_show_reader.py (hunk 8 above)
Comment (sjwhalen) [reply tests/test_dataframe_show_reader.py:240]: This behavior has not changed. If the data type is `int` (whether specified via a data type declaration line or via the optional `default_data_type` param) and the value is not set, we get `ValueError: invalid literal for int() with base 10: ''`.

I have added tests to confirm and document this behavior for both the default data type and the data type declaration line cases. I think this is the safest behavior to prevent swallowing errors from unintentional oversights.

However, I could imagine in the future adding an optional `default_data_value` argument as a shorthand for someone who wants to, for example, initialize a grid to mostly `0` without having to put a `0` in each cell. I'm a little on the fence about adding that shorthand, as it might actually make it harder for someone inheriting the test to read it a glance.

File: dataframe_show_reader/dataframe_assert_equal.py (hunk 2 above)
Comment (jdavenport) [reply dataframe_show_reader/dataframe_assert_equal.py]: Nice.

File: dataframe_show_reader/dataframe_assert_equal.py (hunk 4 above)
Comment (jdavenport) [reply dataframe_show_reader/dataframe_assert_equal.py]: Ahh, good call.

File: tests/test_dataframe_show_reader.py (hunk 8 above)
Comment (jdavenport) [reply tests/test_dataframe_show_reader.py:240]: agreed, i'd rather force the user's hand than make a judgement based on likely intentions

**Review by jdavenport — APPROVED** [review]
Looks good.
